bin/run

```diff
@@ -0,0 +1,163 @@
+# the files passed in the remaining arguments
+function run_shellcheck_against_files() {
+  local shell="${1}"
+  local shellcheck_errors="${2}"
+  shift 2
+
+  files=("$@")
+
+  # Now loop over all scripts, feeding each one to shellcheck and
+  #  collating errors.
+  # Note that we could just collate all of the files into one string to feed
+  #  to shellcheck, but that goes wrong if there are spaces in the path as the
+  #  shell mistakenly splits on WS. There is no way to feed shellcheck the whole
+  #  list as one string separated by NULs, which is the safe way to pass
+  #  filenames.
+  for shell_script in "${files[@]}"; do
+    local shellcheck_status=0
+    local shellcheck_output
+    shellcheck_output="$(shellcheck --shell="${shell}" "${shell_script}")" || shellcheck_status=$?
+
+    if [[ "${shellcheck_status}" -ne 0 ]]; then
+      {
+        printf -- '%s\n\n' "${shellcheck_output}"
+        printf -- '~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\n'
+      }>>"${shellcheck_errors}"
+    fi
+  done
+
+}
+
+source_code="${1:?Source code required}"
+
+# shellcheck disable=SC1091 # sourced file not specified as input
+. /opt/ci-shellcheck-task/lib/ci_configuration.bash
+
+if ! command -V shellcheck >/dev/null 2>&1; then
+  printf "shellcheck not found, so cannot run shellcheck\n"
+  return 1
+else
+  shellcheck_errors=$(mktemp)
+  trap 'rm "${shellcheck_errors}"' EXIT
+
+  # bash_scripts holds list of bash script files found by find commands
+  bash_scripts=()
+
+  # posix_scripts holds list of posix shell script files found by find commands
+  posix_scripts=()
+
+  #regex to match bash shebangs, such as #!/bin/bash, #!/usr/bin/env bash, etc
+  bash_shebang_regex='^#!.*bash'
+  sh_shebang_regex='^^#!.*[ /]sh'
+
+  # find all scripts that are not in a hidden dir or file
+  # loop over the output of the find command, which uses NUL as its separator, hence -d $'\0'
+  while IFS=  read -r -d $'\0' possible_shell_file; do
+    if is_bash_file "${possible_shell_file}"; then
+      bash_scripts+=("${possible_shell_file}")
+    elif is_posix_file "${possible_shell_file}"; then
+      posix_scripts+=("${possible_shell_file}")
+    fi
+  done < <(find "${source_code}" -type f -not -path "${HIDDEN_FILE_OR_DIR_REGEX}" -not -path "${source_code}/${BACKUPS_REGEX}" -not -path "${source_code}/${LOGS_REGEX}" -not -path "${source_code}/${TESTS_REGEX}" -print0)
+
+  # find special case: ci scripts, which are in a hidden dir and so would not
+  # have been found in previous find command, which excludes hidden files/dirs
+  # loop over the output of the find command, which uses NUL as its separator, hence -d $'\0'
+  while IFS=  read -r -d $'\0' possible_shell_file; do
+    if is_bash_file "${possible_shell_file}"; then
+      bash_scripts+=("${possible_shell_file}")
+    fi
+  done < <(find "${source_code}/.ci/bin" "${source_code}/.ci/lib" -type f -print0)
+
+  printf 'Linting %s bash scripts\n' "${#bash_scripts[@]}"
```

Comment:
Why use single quotes for printf?
I use double quotes I am honestly not certain of what difference it makes. I probably knew at some point...
but looking at examples on the Internet and in `man printf`, I see double quotes being used.

Reply:
Preventative

```sh
my_var=123
printf '${my_var}'
printf "${my_var}"
```

would output
```txt
${my_var}
123
```

since `'` does not expand the variables within it. Probably negigible difference but it seems to prevent the misuse of printf.

Reply:
I see. Cool.